File: Calculator/android-native/app/src/main/java/com/qaantum/calculatorhub/calculators/MathCalculators.kt
package com.qaantum.calculatorhub.calculators

import kotlin.math.*

// Percentage Calculator
data class PercentageResult(val value: Double)
class PercentageCalculator {
    fun whatIsXPercentOfY(x: Double, y: Double) = PercentageResult((x / 100) * y)
    fun xIsWhatPercentOfY(x: Double, y: Double) = if (y != 0.0) PercentageResult((x / y) * 100) else PercentageResult(0.0)
    fun percentageChange(from: Double, to: Double) = if (from != 0.0) PercentageResult(((to - from) / from) * 100) else PercentageResult(0.0)
}

// Quadratic Solver
sealed class QuadraticResult {
    data class TwoRealRoots(val x1: Double, val x2: Double) : QuadraticResult()
    data class OneRealRoot(val x: Double) : QuadraticResult()
    data class ComplexRoots(val real: Double, val imaginary: Double) : QuadraticResult()
    object NotQuadratic : QuadraticResult()
}
class QuadraticSolver {
    fun solve(a: Double, b: Double, c: Double): QuadraticResult {
        if (a == 0.0) return QuadraticResult.NotQuadratic
        val d = b * b - 4 * a * c
        return when {
            d > 0 -> QuadraticResult.TwoRealRoots((-b + sqrt(d)) / (2 * a), (-b - sqrt(d)) / (2 * a))
            d == 0.0 -> QuadraticResult.OneRealRoot(-b / (2 * a))
            else -> QuadraticResult.ComplexRoots(-b / (2 * a), sqrt(-d) / (2 * a))
        }
    }
}

// GCD/LCM Calculator
data class GcdLcmResult(val gcd: Long, val lcm: Long)
class GcdLcmCalculator {
    private fun gcd(a: Long, b: Long): Long = if (b == 0L) abs(a) else gcd(b, a % b)
    fun calculate(a: Long, b: Long): GcdLcmResult {
        val g = gcd(a, b)
        val l = if (g != 0L) abs(a * b) / g else 0L
        return GcdLcmResult(g, l)
    }
}

// Factorial Calculator
class FactorialCalculator {
    fun calculate(n: Int): String {
        if (n < 0) return "Error"
        if (n > 170) return "Too large"
        var result = 1.0
        for (i in 2..n) result *= i
        return if (n <= 20) result.toLong().toString() else "%.4e".format(result)
    }
}

// Fibonacci Generator
class FibonacciGenerator {
    fun generate(count: Int): List<Long> {
        if (count <= 0) return emptyList()
        val seq = mutableListOf(0L, 1L)
        for (i in 2 until count) seq.add(seq[i - 1] + seq[i - 2])
        return seq.take(count)
    }
}

// Prime Factorization
class PrimeFactorizationCalculator {
    fun factorize(n: Long): List<Long> {
        if (n <= 1) return emptyList()
        val factors = mutableListOf<Long>()
        var num = n
        var d = 2L
        while (d * d <= num) {
            while (num % d == 0L) { factors.add(d); num /= d }
            d++
        }
        if (num > 1) factors.add(num)
        return factors
    }
}

// Random Number Generator
class RandomNumberGenerator {
    fun generate(min: Int, max: Int, count: Int = 1): List<Int> {
        if (min > max) return emptyList()
        return List(count) { (min..max).random() }
    }
}

File: Calculator/ios-native/CalculatorHub/Calculators/MathCalculators.swift
import Foundation

// Percentage Calculator
struct PercentageResult { let value: Double }
class PercentageCalculator {
    func whatIsXPercentOfY(x: Double, y: Double) -> Double { (x / 100) * y }
    func xIsWhatPercentOfY(x: Double, y: Double) -> Double { y != 0 ? (x / y) * 100 : 0 }
    func percentageChange(from: Double, to: Double) -> Double { from != 0 ? ((to - from) / from) * 100 : 0 }
}

// Quadratic Solver
enum QuadraticResult {
    case twoRealRoots(x1: Double, x2: Double)
    case oneRealRoot(x: Double)
    case complexRoots(real: Double, imaginary: Double)
    case notQuadratic
}
class QuadraticSolver {
    func solve(a: Double, b: Double, c: Double) -> QuadraticResult {
        guard a != 0 else { return .notQuadratic }
        let d = b * b - 4 * a * c
        if d > 0 { return .twoRealRoots(x1: (-b + sqrt(d)) / (2 * a), x2: (-b - sqrt(d)) / (2 * a)) }
        if d == 0 { return .oneRealRoot(x: -b / (2 * a)) }
        return .complexRoots(real: -b / (2 * a), imaginary: sqrt(-d) / (2 * a))
    }
}

// GCD/LCM Calculator
struct GcdLcmResult { let gcd: Int; let lcm: Int }
class GcdLcmCalculator {
    private func gcd(_ a: Int, _ b: Int) -> Int { b == 0 ? abs(a) : gcd(b, a % b) }
    func calculate(_ a: Int, _ b: Int) -> GcdLcmResult {
        let g = gcd(a, b)
        let l = g != 0 ? abs(a * b) / g : 0
        return GcdLcmResult(gcd: g, lcm: l)
    }
}

// Factorial Calculator
class FactorialCalculator {
    func calculate(_ n: Int) -> String {
        guard n >= 0 else { return "Error" }
        guard n <= 170 else { return "Too large" }
        var result: Double = 1
        for i in 2...max(n, 1) { result *= Double(i) }
        return n <= 20 ? String(Int(result)) : String(format: "%.4e", result)
    }
}

// Fibonacci Generator
class FibonacciGenerator {
    func generate(count: Int) -> [Int] {
        guard count > 0 else { return [] }
        var seq = [0, 1]
        for i in 2..<count { seq.append(seq[i-1] + seq[i-2]) }
        return Array(seq.prefix(count))
    }
}

// Prime Factorization
class PrimeFactorizationCalculator {
    func factorize(_ n: Int) -> [Int] {
        guard n > 1 else { return [] }
        var factors: [Int] = []; var num = n; var d = 2
        while d * d <= num { while num % d == 0 { factors.append(d); num /= d }; d += 1 }
        if num > 1 { factors.append(num) }
        return factors
    }
}

// Random Number Generator
class RandomNumberGenerator {
    func generate(min: Int, max: Int, count: Int = 1) -> [Int] {
        guard min <= max else { return [] }
        return (0..<count).map { _ in Int.random(in: min...max) }
    }
}
